### ISSUE TITLE
Getting (wrong-type-argument stringp nil) when invoking navi-search-and-switch

### ISSUE THREAD

**Issue by dakrone** [alphapapa/navi#1]
I opened a buffer and hit `M-s n` and get `(wrong-type-argument stringp nil)`

Here's a backtrace from manually invoking it:

```
Debugger entered--Lisp error: (wrong-type-argument stringp nil)
  regexp-quote(nil)
  navi-search-and-switch()
  eval((navi-search-and-switch) nil)
  pp-eval-expression((navi-search-and-switch))
  funcall-interactively(pp-eval-expression (navi-search-and-switch))
  call-interactively(pp-eval-expression nil nil)
  command-execute(pp-eval-expression)
```

**Comment by alphapapa** [comment]
This appears to be because `outline-promotion-headings` is nil.  What is the major mode of the buffer you're running this command in?

**Comment by dakrone** [comment]
> What is the major mode of the buffer you're running this command in?

It was a `java-mode` file.

**Comment by alphapapa** [comment]
I checked a `java-mode` buffer and `outline-promotion-headings` was set to:
```el
(("// * " . 1)
 ("// ** " . 2)
 ("// *** " . 3)
 ("// **** " . 4)
 ("// ***** " . 5)
 ("// ****** " . 6)
 ("// ******* " . 7)
 ("// ******** " . 8))
```
Do you have the latest version of `outshine` installed?  If so, please check the value of `outline-regexp` in a `java-mode` buffer.  Thanks.

**Comment by dakrone** [comment]
The value of `outline-regexp` in the java-mode buffer is: 

```
"[^#\n
]"
```
(or `"[^#\n^M]"` without translating the `^M` to a newline)

**Comment by dakrone** [comment]
> Do you have the latest version of outshine installed?

Sorry, forgot to mention, I'm using the latest version of outshine from MELPA, `outshine-20170414.1217`

**Comment by alphapapa** [comment]
I don't understand that `outline-regexp` value.  Here's mine in a `java-mode` buffer:
```
outline-regexp is a variable defined in ‘outline.el’.
Its value is "// [*]\\{1,8\\} "
Local in buffer charlie-d02fb.java; global value is "[*\f]+"
```
Have you customized it or changed it in some way?

**Comment by dakrone** [comment]
> Have you customized it or changed it in some way?

Nope, I've never even used outshine before, is this perhaps something that `semantic-mode` could be interfering with?

I'm happy to do

```emacs-lisp
(add-hook 'java-mode-hook (lambda () (setq-local outline-regexp "// [*]\\{1,8\\} ")))
```

If that would help, I just did not expect that this would need to be set

**Comment by dakrone** [comment]
I just tried this again, interestingly, if I eval `outshine-regexp` I get `"// [*]\\{1,8\\} "` (correct), however, when I try to run `navi-search-and-switch`, I see the following messages in the `*Messages*` buffer:

```
Searched 1 buffer; no matches for "// \* "
helm-M-x: Wrong type argument: stringp, nil
```

It looks like it's still searching for the wrong thing?

**Comment by alphapapa** [comment]
> Nope, I've never even used outshine before, is this perhaps something that semantic-mode could be interfering with?

I don't think so.

> I just did not expect that this would need to be set

You're right, it shouldn't be necessary.

>  if I eval outshine-regexp I get "// [*]\\{1,8\\} " (correct)

Did you mean `outline-regexp`?  `outshine-regexp` doesn't exist, AFAIK.  (It does start to get confusing.)

> It looks like it's still searching for the wrong thing?

Well, this line:
```
Searched 1 buffer; no matches for "// \* "
```
Would seem to indicate that the search worked, but there are no headings in the file.  Is that correct?

This line:
```
helm-M-x: Wrong type argument: stringp, nil
```
has no backtrace, but it does sound like the same problem.  Did you run this in a `java-mode` buffer too?

Which version of Emacs are you using?

**Comment by dakrone** [comment]
> Did you mean outline-regexp? outshine-regexp doesn't exist, AFAIK. (It does start to get confusing.)

Yep, sorry for the typo.

> This line ... has no backtrace, but it does sound like the same problem. Did you run this in a java-mode buffer too?

This is the same error/backtrace as in the initial issue message

> Would seem to indicate that the search worked, but there are no headings in the file. Is that correct?

Oh man, I think I understand the problem now, I did not realize that `outline-mode` was searching for headings in the org-mode format, I naively assumed it would automatically use things like function and class names as headings (I have never used outline-mode before). When I add a headline by doing `// * Example heading` in the file, it correctly detects it and `navi-search-and-switch` works correctly!

I think this can be closed now, sorry for the roundabout debugging.

One more suggestion though, maybe `navi-search-and-switch` (or outshine) can catch the `no matches for ...` error (when `nil` is passed instead of a list of matches) and log a messages like:

```
No headings were found in the file, searched for "// \* "
```

That way it would be clear to people that navi-mode isn't broken, they just don't have any headlines in the file?

**Comment by alphapapa** [comment]
Ok, that explains it.  :)

I think what you want is the "keyword" functionality in `navi`.  Try running `outshine-navi` and then pressing `f`.

BTW, you may also find the `helm-navi` package useful for this, especially the `helm-navi-headings-and-keywords-current-buffer` command.
